maup/repair.py

```diff
@@ -1,27 +1,1217 @@
+
+                shared_perimeters = []
+
+                for b_ind in this_hole_boundaries_df.index:
+                    g_ind = this_hole_boundaries_df["target"][b_ind]
+                    shared_perimeters.append(
+                        (
+                            b_ind,
+                            (this_hole.boundary)
+                            .intersection(geometries_df["geometry"][g_ind].boundary)
+                            .length,
+                        )
+                    )
+
+                max_shared_perim = sorted(shared_perimeters, key=lambda tup: tup[1])[-1]
+                poly_to_add_to = this_hole_boundaries_df["target"][max_shared_perim[0]]
+
+                geometries_df["geometry"][poly_to_add_to] = unary_union(
+                    [geometries_df["geometry"][poly_to_add_to], this_hole]
+                )
+                holes_df = holes_df.drop(h_ind)
+
+                activity_this_round = True
+
+            else:  # Partially fill the hole!
+
+                (
+                    poly_to_add_to,
+                    piece_to_connect,
+                    new_holes_this_time,
+                ) = partial_fill_data(this_hole, this_hole_boundaries_df)
+
+                if poly_to_add_to is not None:
+                    # If a relative threshold is set, check that the piece to connect is
+                    # under the threshold relative to poly_to_add_to before adjoining it
+                    # to the polygon.
+                    geometries_df["geometry"][poly_to_add_to] = unary_union(
+                        [geometries_df["geometry"][poly_to_add_to], piece_to_connect]
+                    )
+                    holes_df = holes_df.drop(h_ind)
 
-class AreaCroppingWarning(UserWarning):
-    pass
+                for k in range(len(new_holes_this_time)):
+                    new_holes.append(new_holes_this_time[k])
+
+                    activity_this_round = True
+
+        if activity_this_round == True:  # Add any new holes for the next round!
+            for k in range(len(new_holes)):
+
+                hole_to_append_df = geopandas.GeoDataFrame(
+                    columns=["polygon indices"],
+                    geometry=geopandas.GeoSeries([new_holes[k]]),
+                    crs=geometries_df.crs,
+                )
+                hole_to_append_df["polygon indices"][0] = set()
+
+                holes_df = pandas.concat(
+                    [holes_df, hole_to_append_df], ignore_index=True
+                )
+                holes_df.reset_index(drop=True)
+
+        else:  # All done!
+            fill_complete = True
+
+    return geometries_df
+
+
+# Partially fill a hole with 4 or more boundary components by attaching two
+# non-adjacent boundary components, leaving behind two new holes, each with strictly
+# fewer boundary components than the original.
+
+
+def partial_fill_data(hole0, hole_boundaries_df0):
+
```

Comment:
Here's how this might be converted to a docstring:
```suggestion
def partial_fill_data(hole0, hole_boundaries_df0):
    """
    Partially fill a hole with 4 or more boundary components by
    attaching two non-adjacent boundary components, leaving behind
    two new holes, each with strictly fewer boundary components than
     the original.
    """
```